File: src/algorithms/hash-maps/tracking/intersection-of-two-arrays/sources/IntersectionOfTwoArrays.java
// Intersection of Two Arrays — find common elements using a hash set
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

public class IntersectionOfTwoArrays {
    public static List<Integer> intersectionOfTwoArrays(int[] numbersA, int[] numbersB) {
        HashSet<Integer> setA = new HashSet<>(); // @step:initialize
        for (int elementIndex = 0; elementIndex < numbersA.length; elementIndex++) {
            setA.add(numbersA[elementIndex]); // @step:insert-key
        }
        List<Integer> result = new ArrayList<>();
        for (int elementIndex = 0; elementIndex < numbersB.length; elementIndex++) {
            int currentNum = numbersB[elementIndex];
            if (setA.contains(currentNum)) { // @step:lookup-key
                result.add(currentNum); // @step:key-found
                setA.remove(currentNum);
            }
        }
        return result; // @step:complete
    }
}

File: src/algorithms/hash-maps/tracking/intersection-of-two-arrays/sources/intersection-of-two-arrays.py
# Intersection of Two Arrays — find common elements using a hash set
def intersection_of_two_arrays(numbers_a, numbers_b):
    set_a = set()  # @step:initialize
    for element_index in range(len(numbers_a)):
        set_a.add(numbers_a[element_index])  # @step:insert-key
    result = []
    for element_index in range(len(numbers_b)):
        current_num = numbers_b[element_index]
        if current_num in set_a:  # @step:lookup-key
            result.append(current_num)  # @step:key-found
            set_a.discard(current_num)
    return result  # @step:complete
